### ISSUE TITLE
[DX] Switch to a simplified array syntax for database connection information in settings.php

### ISSUE THREAD

**Issue by olafgrabienski** [backdrop/backdrop-issues#2231]
I just tried to install Backdrop 1.5 in a local AMPPS environment. During installation, after being on the language page, I got the message

`Fatal error: Class 'DatabaseTasks_' not found in {...}/core/includes/install.inc on line 1362`

and wasn't able to continue the installation.

Line 1362 concerns the database driver. Googling something like _drupal error database driver_ led me to a [comment on d.org regarding Backup and Migrate](https://www.drupal.org/node/1999010#comment-7841581) about a problem with special characters in the database password string. I had a look in my _settings.php_, noticed a forward slash in the password string, changed the database password in the relevant places and was able to continue the installation.

Are forward slashes (and possibly particular other special characters) in the database password string not supported by Backdrop for a certain technical reason? If so, we should document it somewhere. If not, can we fix that?

ADVOCATE: @yorkshire-pudding

**Comment by Dinornis** [comment]
I experienced a similar issue with a password that included special characters, the only difference was that I did not see an error message, just a blank page during install.  Not sure what would be the best way to address this.

**Comment by olafgrabienski** [comment]
Today I got the same, or very similar, issue again. Installation of Backdrop failed, displaying the general message "The website encountered an unexpected error. Please try again later."

When I went back to the site URL without the installer related query strings, I got a more specific notice:

`Notice: Undefined index: driver in {...}/core/includes/database/database.inc on line 1718`

As in the first post above, it looks again like a database driver related issue. (Maybe sort of misleading, because the driver is the first element of `$database` in `settings.php`.) I was however able to install Backdrop with a new database using a new password, and I guess that worked because there were no 'problematic' characters in the string.

So I ask again: are particular special characters not supported in the database password string?

**Comment by LeeteqXV** [comment]
I just ran into this on a test installation of 1.6.3, where the password contained a semicolon (;) and a hash (#) character on two different/adjacent locations in the password. I replaced those two with underscores (so that at least it had 1 special character in it), reloaded the error page, and the installation continued without any further problems.

**Comment by olafgrabienski** [comment]
Documenting for testing purposes: characters mentioned in the posts above that might be problematic:

- forward slash: `/`
- semicolon: `;`
- hash sign: `#`

Feel free to add more.

**Comment by daggerhart** [comment]
This is because the [bootstrap is using parse_url()](https://github.com/backdrop/backdrop/blob/1.x/core/includes/bootstrap.inc#L776) to load this string into an array, so anything that is part of the url schema is problematic.  Imo, this is one of the benefits to the array structure D7 adopted, it allowed for better passwords w/ fewer parsing errors. 

Maybe we can allow the array syntax through? Something like:

```php
// if an array and contains 'database' key, then this is likely a connection-ready array
if (is_array($database) && isset($database['database'])){
  $databases['default']['default'] = $database;
}
```

Alternatively, we could replace the password with something safe prior to parsing the url, then re-instate it afterwords directly into the $databases array.

Scratch

```php
$database = "mysql://user:paasd!$%@[]:;%#$]word@localhost/database_name";

// get the password separate from the rest of url
$tmp = explode('@',$database);
array_pop($tmp); // remove hostname+
$tmp = implode('@', $tmp); // join on @ to get string as everything but hostname
$tmp = explode(':', $tmp); // split on colon to remove scheme and username
array_shift($tmp); // scheme~ish
array_shift($tmp); // username~ish
$password = implode(':', $tmp); // remaining is the password

$database = str_replace($password, 'safe-placeholder', $database);
// ... later
$databases['default']['default']['password'] = $password;
```

That ^ should be safe every time, though very ugly.

**Comment by olafgrabienski** [comment]
@daggerhart Do you suggest to change `settings.php`, so that the default would be to use the array format instead of the simple database configuration?

**Comment by laryn** [comment]
I just hit this using an autogenerated password. It was a big pain to sort out as things had already been partially installed and I had to:

- reset the password for the database
- edit settings.php to update the password there
- reload and still got errors so I decided to start fresh and replaced settings.php with the default version of that file
- began to reinstall then got a fatal error because config directories weren't empty
- emptied those, reinstalled and now I'm getting `Call to undefined function field_attach_load()` so I'm planning to just nuke the whole thing and start fresh

**Comment by daggerhart** [comment]
@olafgrabienski sorry, i missed your message to me. 

Yes, if we change to using the array format during settings.php generation, we'll avoid users running into issues like @laryn just experienced.

I don't know of any great reason to using the single string/url format. As a user, I think it is much harder to understand because you have to be somewhat familiar with authentication based url syntax and avoid its special characters.

**Comment by olafgrabienski** [comment]
> I don't know of any great reason to using the single string/url format [for database configuration in settings.php].

@quicksketch @jenlampton To fix errors during installation due to special characters in the database password: Do you think it's possible to change the database configuration in `settings.php*, so that it uses the array format instead of the one line format? Or is there a particular reason why Backdrop has to use the one line format?

**Comment by jenlampton** [comment]
> Or is there a particular reason why Backdrop has to use the single string format?

@olafgrabienski backdrop does not need to use the single string format, the array format from Drupal 7 will work as well.

**Comment by olafgrabienski** [comment]
Thanks @jenlampton ! To change `settings.php` in such an essential part seems a big change to me. Does it make sense, and are there caveats? I'd appreciate some more feedback from experienced Backdrop developers before we make a PR!

**Comment by jenlampton** [comment]
There's no need for a PR, backdrop already works this way.

On your site: copy the array syntax from a D7 settings.php file and see if
that fixes the problem. If so we can close this issue.

**Comment by laryn** [comment]
@jenlampton If we want it to use the array syntax by default on new installs (to avoid the special characters in the MySQL password causing a problem) we'd at least need a PR to change the default `settings.php` file, wouldn't we? And maybe some fixes where Backdrop saves settings during installation?

**Comment by jenlampton** [comment]
We don't want to use the ugly array syntax by default.

The simper syntax was chosen because all people will understand it
instantly. It's a universal syntax, and will benefit 100% of backdrop
users. For the minority of sites that still need to use the array syntax
for any of the various edge-cases, that Is an option too.

**Comment by laryn** [comment]
@jenlampton Can you think of another solution to the special character issue caused by the single line syntax? (ie. what this thread is about)

I don't think I'd call that an 'edge case' and expect new users who run into this to be able to figure out why they can't even install Backdrop in the first place. For example, using the cPanel "generate password" functionality on a new MySQL database user will almost always include a special character like this.

**Comment by jenlampton** [comment]
It's important that we avoid using a crazy backdrop-specific database connection syntax (by default) because that's the first thing people will see in settings.php. New users will see that first when setting up Backdrop, and that creates a poor first impression too.

Another thing we should do is make the installer UI be smarter about special characters in the database password. If a user enters one into the UI during installation, the installer should switch to using the crazy syntax when it writes to settings.php, or escape special characters (like slashes) when necessary. At the very least, it could throw an error or warning message letting people know what's up rather than just crashing the site. :(

Since what we're using now is a universal syntax, maybe we could look at what other all the other systems are doing about it, for reference?

**Comment by laryn** [comment]
I'm not sure what's so bad about the array format (I actually think it's clearer). How often will the 80% need to open `settings.php` anyway following a successful installation via the browser interface?

That said, I don't really have strong opinions as long as the issue gets solved.

**Comment by klonos** [comment]
Sorry @jenlampton, but I'm with the rest of the people on this one...

1. I find that the array syntax is more self-explanatory than the URL syntax. Mainly because it uses placeholders for the parts in a way that resembles field label/input pairs. IMO this would help beginner devs more.
2. The array syntax eliminates this problem here, which is a showstopper and causes a very bad UX for beginners trying to install Backdrop for the first time. Especially in very common environments like cPanel.
3. "*Backdrop values site builders over coders*". Very few will need to directly mess with the settings.php file (which is code) in the first place. Or at least that's how things should be. Things are meant to be handled by our installer UI.

Now, if we can find an easy way to make the password field smart enough to detect and escape problematic characters before writing to the settings.php file, then I'm equally OK with that.

**Comment by klonos** [comment]
...to sum things up:

> It's important that we avoid using a crazy backdrop-specific database connection syntax (by default)...

I agree.

> ...because that's the first thing people will see in settings.php.

These people will most likely be developers. Site builders do not mess with the settings.php file. At least beginner site builders that expect things to work from the UI.

> New users will see that first when setting up Backdrop, and that creates a poor first impression too.

Not poorer impression than a "broken" installer that "doesn't work" and that leads you to have to fiddle with code.

**Comment by docwilmot** [comment]
Agreed.
We need to either default the array syntax, or find/build an alternative to `parse_url()`.

**Comment by jenlampton** [comment]
1) Top priority: If we're talking about the site-builder experience then the format we use in the settings.php file is irrelevant, since the site-builder won't be looking at it. We should fix the installer to either throw a validation error if a password is entered that won't work, or, ideally, deal properly with whatever password is provided. It's not a good idea to solve UX problems by creating DX problems when it's not necessary to do so.

2) Our second priority is for the beginner coder (or beginner Backdrop user), and for them it's very important that we not use the array syntax from D7. I have a feeling all of you are too advanced to understand what it feels like for a non-developer to look at settings.php in Drupal 7 for the first time. I've trained hundreds of people on Drupal. The settings.php file shatters people's confidence - it deters them from thinking that they can do anything at all with Drupal's code. 

> I'm not sure what's so bad about the array format

* There are 187 lines of documentation needed to explain this syntax. Starting right off with 187 lines of "help" on how to hook up the database says "This software is so complicated that you can't even understand the first step without reading this epic". That is not the first impression we want to give people!
* Nested arrays are horribly confusing to people who don't know what arrays are, or understand array syntax in PHP.
* Why is the first key "default" when you only have one database? Why is the second key also "default" when you still only have one database? These are unnecessary abstractions for the majority of users who only have a single database. In order to use the existing array syntax we require that people understand concepts that are irrelevant to the their own use case. That's not in line with backdrop's philosophy.

I think we may are be missing the point here. Changing the syntax back to the mess from D7 is NOT the answer, and we should stop talking about doing that by default as it would be a huge mistake. We changed it for good reasons and we should not reopen that conversation.

If we want to fix the UX of the installer, fine, let's do that.

If we want to improve the DX of the settings.php file, fine, let's do that too! But let's not move backwards and throw our principles out the window while doing so. Let's find a new solution that doesn't have all the problems we've already solved.

**Comment by docwilmot** [comment]
@jenlampton I think the 187 lines of documentation still exist; we just moved them to our docs site. :smile: 
Here's the current settings for dbases:
```
/**
 * Database configuration:
 *
 * Most sites can configure their database by entering the connection string
 * below. If using master/slave databases or multiple connections, see the
 * advanced database documentation at
 * https://api.backdropcms.org/database-configuration
 */
$database = 'mysql://user:pass@localhost/database_name';
$database_prefix = '';
```

Here's what it would look like:

```
/**
 * Database configuration:
 *
 * Most sites can configure their database by entering the connection details
 * below. If using master/slave databases or multiple connections, see the
 * advanced database documentation at
 * https://api.backdropcms.org/database-configuration
 */
  $database = array(
    'driver' => 'mysql',
    'database' => 'databasename',
    'username' => 'username',
    'password' => 'password',
    'host' => 'localhost',
    'prefix' => '',
  );
```

I dearsay I think the latter is certainly less scary.

In either case if we continue to use the string format, we must concede that users will have issues with password generators OR they will get nasty exceptions. Or we come up with an alternative to `parse_url()`. Frankly I'm not seeing many other options.

**Comment by jenlampton** [comment]
>  I think the 187 lines of documentation still exist; we just moved them to our docs site. 

yes, because they aren't necessary for people reading the settings.php file, thanks to our new cleaner syntax ;)

> Frankly I'm not seeing many other options.

That's why there's more than one of us!  Here's what WP does: https://codex.wordpress.org/Editing_wp-config.php

**Comment by docwilmot** [comment]
> I gave two examples above of how this can be avoided

Sorry missing those, can you explain? Solutions are always a good thing.

**Comment by jenlampton** [comment]
> Here's what it would look like:

```
/**
 * Database configuration:
 *
 * Most sites can configure their database by entering the connection details
 * below. If using master/slave databases or multiple connections, see the
 * advanced database documentation at
 * https://api.backdropcms.org/database-configuration
 */
  $database = array(
    'driver' => 'mysql',
    'database' => 'databasename',
    'username' => 'username',
    'password' => 'password',
    'host' => 'localhost',
    'prefix' => '',
  );
```

Actually, this looks pretty clean and tidy... Does this work for backdrop? Can we test it real quick?

**Comment by jenlampton** [comment]
> Sorry missing those, can you explain?

If a user enters a password with special characters into the UI during installation, the installer should 1) switch to using the array syntax when it writes that info into settings.php, or
2) properly escape special characters (like slashes) when necessary.

**Comment by docwilmot** [comment]
> If a user enters a password with special characters into the UI during installation, the installer should 1) switch to using the array syntax when it writes that info into settings.php, or
2) properly escape special characters (like slashes) when necessary.

Makes sense

>Actually, this looks pretty clean and tidy... Does this work for backdrop? Can we test it real quick?

Shouldnt work; would need a slight adjustment to `bootstrap.inc`.

**Comment by jenlampton** [comment]
> Shouldn't work; would need a slight adjustment to bootstrap.inc.

I thought for a sec that was an alternate Drupal 7 syntax and I was like "How have I never seen this before?!?!" I think that is a good recommendation @docwilmot and I don't think new users will have trouble with that syntax :)

**Comment by laryn** [comment]
Looks good to me, @docwilmot!

**Comment by olafgrabienski** [comment]
Nice! @docwilmot, would you be willing to make a PR including the `bootstrap.inc` adjustment?

**Comment by klonos** [comment]
@jenlampton sorry, I understand now that when you say "scary" or "crazy" you are referring to the `$databases['default']['default'] = ` part while I thought you were referring to the array element syntax. When I was saying "self-explanatory", I was referring to this part:

```
    'driver' => 'mysql',
    'database' => 'databasename',
    'username' => 'username',
    'password' => 'password',
    'host' => 'localhost',
    'prefix' => '',
```

So, basically yes, what @docwilmot suggests 😄 

...or this:

> If a user enters a password with special characters into the UI during installation, the installer should ... properly escape special characters (like slashes) when necessary.

...which is the perfect solution for both the installer problem and the better beginner DX when first facing settings.php.

**Comment by jlfranklin** [comment]
Will existing `$databases['default']['default']` continue to work after the PR?  Otherwise, this has the potential to break a site on upgrade.

**Comment by jenlampton** [comment]
> Will existing $databases['default']['default'] continue to work after the PR? Otherwise, this has the potential to break a site on upgrade.

We should aim to keep the old syntaxes working too, if possible. We can mark them as deprecated and remove in 2.0 if we want.

**Comment by docwilmot** [comment]
>If a user enters a password with special characters into the UI during installation, the installer should ... properly escape special characters (like slashes) when necessary.
...which is the perfect solution for both the installer problem and the better beginner DX when first facing settings.php.

This would still break if someone entered manually into `settings.php` using the string format. Its a reasonable option, but I dont like the idea that the start of an install could be that fragile. I think a simple array solves everyone's problems.

We'd need to ensure that all of 
- the simple string
- the simple array
- and the complex `$databases['default']['default']` 

still work. We cant get rid of the complex one even in 2.0 unless we commit to removing slave db support. So at best we can deprecate the simple string now.

**Comment by jenlampton** [comment]
> We cant get rid of the complex one even in 2.0 unless we commit to removing slave db support

...or rework how slave databases are added, too. Either way that is certainly an edgecase. These other issues mentioned here are more important as they are likely to come up often. I like the new array syntax @docwilmot proposed, and if we start by replacing the string syntax and deprecating that I think it's still a win!

**Comment by klonos** [comment]
> This would still break if someone entered manually into `settings.php` using the string format.

We could then add something along the lines of "the following characters need to be escaped" in the comments above the db connection line.

**Comment by simone960** [comment]
I'm still encountering the issue when installing 1.8.0. I think this should be fixed whichever way needed. The beginners will get stuck right in the beginning when installing the backrop for the first time and need to google around, read the discussion and get the fix ==" I'm not a developer. What I fix is just to reset the password and removed those special character. It took me quite a while to get the site installed. This is not really not impressive...

**Comment by jenlampton** [comment]
Thanks for your feedback @simone960. I agree, we should get this sorted out asap. Tagging for the 1.8.1 release.

**Comment by simone960** [comment]
Glad to hear that @jenlampton. Thanks for listening and making a decisive move. The door has to be first opened to people before we can talk about how great the interior design is.

**Comment by jlfranklin** [comment]
FYI, PHP handled this better up to 7.0.12 and 5.6.27.

https://3v4l.org/Urngf

**Comment by jlfranklin** [comment]
I think there are three things we should do, each can get its own issue:

- [ ] Properly encode/decode parts of the database URL to support special characters (#2904)
- [ ] Better error reporting if the database URL is unparsable or missing critical fields.
- [ ] Support the simplified array format

**Comment by olafgrabienski** [comment]
@jlfranklin Wouldn't #2904 be sufficient to fix the actual bug?

**Comment by jlfranklin** [comment]
Only if you are entering the username and password for the database via Backrdrop's installer.  If you type it directly into `settings.php`, then no, #2904 won't fix it.

**Comment by olafgrabienski** [comment]
@jlfranklin Thanks for your feedback. In the meantime I had a look at the PR for #2904 , looks good to me! Unfortunately I don't know how to help with the other two sub issues. If we don't fix them in time, I'd suggest however to merge #2904 as soon as possible because I assume that using Backdrop's installer is the default installation method for beginners.

**Comment by quicksketch** [comment]
I've merged @jlfranklin's recommendations in #2904 and #2914, so we should no longer have problems with the string format when installing via install.php.

I've always thought the string format was simpler/easier, but the change to PHP's `parse_url()` in new versions of PHP will really put a negative impact on the ease of use if URL-encoding is required. Encoding strings isn't something I would expect anyone to be able to do by hand, which is what the goal of the string format in the first place.

As we've largely fixed the bugs through @jlfranklin's PRs, let's retitle this to a DX feature request to switch to a simplified array syntax instead (if we want to go down that route).

**Comment by yorkshire-pudding** [comment]
Any ideas, what needs to happen where for the simplified array format to be supported in `settings.php` and `settings.local.php`? I've had to resort to the D7 method so I can put predefined credentials with complex passwords in for deploying to different environments.

**Comment by yorkshire-pudding** [comment]
This issue seems to be the cause of another issue in the forum - https://forum.backdropcms.org/forum/deployment-problems

**Comment by olafgrabienski** [comment]
Not sure if it was *exactly* this issue, but I also recently ran again in a situation as described in the forum: Wasn't able to get a site running on another server, and only because I *remembered* issues with the database password, I suggested to the hoster to please change the password, and then the site ran smoothly. This must be horrible for newcomers. If we can avoud such a situation by fixing this issue, so please let's fix it. (Happy to help, but don't know how.)

**Comment by klonos** [comment]
@olafgrabienski the technical direction here has been outlined in https://github.com/backdrop/backdrop-issues/issues/2231#issuecomment-341854268 and https://github.com/backdrop/backdrop-issues/issues/2231#issuecomment-341938604

Having re-read through the entire thread, I would sum things up to the following goals:
1. we want to support the single-line string format:
   ```php
   $database = 'mysql://user:pass@localhost/database_name';
   $database_prefix = '';
   ```
   Ideally, we'll add an inline comment in `settings.php` to be cautioning people editing that setting directly in the file that there are certain risks with using some special characters with that format.
2. we also want to support a simplified array format:
   ```php
   $database = array(
     'driver' => 'mysql',
     'database' => 'databasename',
     'username' => 'username',
     'password' => 'password',
     'host' => 'localhost',
     'prefix' => '',
   );
   ```
3. we also want to keep supporting the advanced array format, which allows specifying multiple databases (notice the plural `$databases` instead of singular `$database` in the variable name):
   ```php
   $databases['default']['default'] = array(
     'driver' => 'mysql',
     'database' => 'databasename',
     'username' => 'username',
     'password' => 'password',
     'host' => 'localhost',
     'prefix' => '',
   );
   ```
4. we want to do the above without breaking backwards compatibility with settings copied over from D7 `settings.php` files, and with placing as many "safeguards" as possible, in order to prevent misconfigured db arrays or strings that cannot be parsed (some of these things are already in place with previous commits in other issues).

**Comment by muranod** [comment]
That was me with the deployment problems. I'm actually relieved to find out that it might have been the special characters in the database password that might have been causing my issues (over several days of trying to fix it).  I'm not a programmer or developer - just building my own site from d7, d8, d9 experience. The array format would have been easier in my case, simply because of its clarity.

**Comment by yorkshire-pudding** [comment]
I've added a PR for this issue.
https://github.com/backdrop/backdrop/pull/4567

Tugboat not working - I can close and reopen later if issue cleared up.

I've tested locally and it works with simplified string, this new simplified array and still works with old D7 style array.
If testing locally in something like Lando, DDEV or Docksal, you may need to add this line to your `settings.php`

```php
if (isset($_SERVER["BACKDROP_SETTINGS"])) unset($_SERVER["BACKDROP_SETTINGS"]);
```

**Comment by yorkshire-pudding** [comment]
Needs a change in `backdrop_rewrite_settings()` within `install.inc` but not sure how. Any ideas?

**Comment by klonos** [comment]
@yorkshire-pudding re the last comment, it seems that `backdrop_rewrite_settings()` is expecting all settings in `settings.php` to be single lines, so in the case of
```php
$database = array (
  'driver' => 'mysql',
  'database' => 'database_name',
  'username' => 'user',
  'password' => 'pass',
  'host' => 'localhost',
  'prefix' => '',
);
```

...it only replaces the `$database = array (` bit and ignores the rest of the lines that belong to the same array declaration.

I'm working on other things now, but I can have a look later (unless someone else beats me to it).

**Comment by yorkshire-pudding** [comment]
@klonos - got an idea - just testing it.

**Comment by yorkshire-pudding** [comment]
Ok. I've got so install works and addressed most of the feedback. However, tests are failing, and I think that is because one or more tests needs to be rewritten.

Can anyone point to which ones or where I even start looking?

**Comment by yorkshire-pudding** [comment]
Install runs, tests pass; by which of course I mean all tests pass except for `testLayoutUpgrade()` and cspell. The cspell issues are nowhere near what I've worked on.

**Comment by yorkshire-pudding** [comment]
@klonos - you mentioned on the PR:
> Lets also mention which of these keys are optional, and what they default to if omitted.

As far as I can tell, `port` is optional but that is already made clear in the comment.
Looking at the code before, if `prefix` is left blank it reverts to `''` but we already have that anyway.

Interestingly, the code that checks what is added only says:
```php
    $required_parts = array('driver', 'database', 'username', 'host');
```

I don't know why `password` wouldn't be required, but I wouldn't want to say it was optional as the vast majority will have a database password.

If there is anything I'm missing, please let me know.

**Comment by yorkshire-pudding** [comment]
I've seen on functions the concept of optional and default, but I'm really not clear what we want here. What sort of format do we want to use. Keep in mind that there was no guidance before and there are no models for this elsewhere in settings.

This is what the default screen looks like:
![image](https://github.com/backdrop/backdrop-issues/assets/20357797/016a7f26-6e42-4943-ab90-bd339171be71)

So we do seem to default in the UI installer to `127.0.0.1` (see #2520) but from what I understand `localhost` is better if it works as it uses a socket rather than TCP/IP so is faster.

Required:
- driver
- database
- username
- host (this was a default of `localhost` in `settings.php` but is a default of `127.0.0.1` in the UI installer)

Optional
- port - normally standard (i.e. `3306) - this is the only one not actually included in the current example.
- prefix - normally `''` and is set by default now, so I don't think we should leave it out. 
- password - I guess defaults to `''` but I'm not that keen on suggesting that is optional

**Comment by klonos** [comment]
Yeah, we allow adding that information either manually/directly in the settings.php file, or via the installer. It's not good that we have different defaults between the two - there should be consistency.

Also, `driver` will always be `mysql`. We don't expose this in the UI, so I think that we can omit it in the array in settings.php. We should still allow for it to be optionally specified I guess, but that would mean that the site is an instance of Silkscreen - not Backdrop. Right?

So perhaps we should simplify the array to this:
```php
$database = array(
  'database' => 'database_name',
  'username' => 'user',
  'password' => 'pass',
  'host' => 'localhost',
);
```

And then in the comments note that there are other optional keys allowed, and what each one of them is used for.

**Comment by yorkshire-pudding** [comment]
@klonos - I've updated with guidance and reduced default required setting defaults as agreed.

Please could I ask for your support on https://github.com/backdrop/backdrop/pull/4567#discussion_r1394631098
It might be nice, if you have time, to do a video call where you can teach me, but if not, please can you suggest some changes.
Thanks

**Comment by yorkshire-pudding** [comment]
@izmeez - there is a PR - look at the top, but it is not robust enough to merge; it is currently blocked by https://github.com/backdrop/backdrop-issues/issues/6297; hopefully @klonos will have some time to work on that issue soon.

**Comment by bugfolder** [comment]
I tend to agree that the "Database configuration" comment in settings.php needs some updating; at the very least, the URL of "advanced database documentation" should be updated to the docs.backdropcms.org URL (the current api.backdropcms.org redirects there). But I do think that should be a separate issue, something like "Improve the database configuration comments in settings.php."

So this: https://github.com/backdrop/backdrop-issues/issues/6378

**Comment by yorkshire-pudding** [comment]
This was discussed at the Weekly Dev Meeting on 2024-09-05. See https://www.youtube.com/live/s9Y-wZLSWpM?si=EErsKR8C25QopmzB&t=2678

Key points:

- Should not be blocked by the desire to have a better method for replacing text in `settings.php` files so proceed.
- Minimise documentation in `settings.php` and put rest into documentation.

**Comment by yorkshire-pudding** [comment]
The PR has been rebased. I've shortened the comment to:

```php
/**
 * Database configuration:
 *
 * Most sites can configure their database by entering the connection details
 * below. If you need to:
 *   - customize the 'port', 'prefix', or 'driver'
 *   - use primary/replica databases
 *   - use multiple connections
 * see the advanced database documentation at
 * https://docs.backdropcms.org/database-configuration
 */
```

To recap, this includes:
- changes to settings.php to give a much simplified database array
- change bootstrap to support this format as well as the string format and the D7 array format
- change the installer to write database setting in this format.  It uses a line count to set how many lines to remove based on the default array being in settings.php.

@quicksketch @jenlampton @laryn @docwilmot 
Happy for any feedback.
I can start to look at documentation changes if we're agreed mostly on the approach.

**Comment by quicksketch** [comment]
We discussed this today in the weekly meeting. Some changes discussed and agreed upon (both @yorkshire-pudding and @izmeez were present):

- Add charset and collation as allowed options in the new syntax, but provide reasonable defaults.
- Include charset and collation in the docblock list of advanced configuration options which can be set but link to the documentation for details.
- The default for charset should be changed to `utf8mb4`. So if you're using the simplified array syntax, that charset is implied.

**Comment by yorkshire-pudding** [comment]
@izmeez - by all means open a ticket for discussion on `collation`

In `/core/includes/database/charset_converter.inc` a default for `collation` is set of `utf8mb4_general_ci` so it makes sense to be that, at least for now.
Also in `utf8mb4IsSupported()` within `/core/includes/database/mysql/database.inc` it is used

**Comment by yorkshire-pudding** [comment]
@quicksketch @laryn @docwilmot @jenlampton 
I have updated the PR.  It is generating a failure for PHP 5.6 batch 1 but I don't think this is related.:

```
Detailed test results
---------------------
---- Image: Image style flood protection (ImageStyleFloodProtection) ----

Status    Group      Filename          Line    Function                            
------------------------------------------------------------------------------------------------------------------------
Fail      Other      image.test        1970    testFloodProtection()                                                   
    Correct number of images (1) return 403 access denied.
Error: Process completed with exit code 1.
```

@quicksketch - I did also have a look at `/core/scripts/install.sh` and thought it may make sense to move that on from using a db url too.  As `bee` already collects the information separately it would be relatively simple to update that, but I don't know if anything else depends on the current setup.
Also, I did try to find if any tests needed updating but I wasn't able to find it if there were (while I'm familiar with the `bee` testing regime, I am less familiar with Backdrop's testing approach).

**Comment by yorkshire-pudding** [comment]
I've updated the `/.tugboat/settings.local.php` to use a `$database` array, but it still didn't appear to be working, so I have made `backdrop_install_config_location()` more flexible so it can cope with either a string or an array. Now tugboat is working and showing that `utf8mb4` is enabled:
![image](https://github.com/user-attachments/assets/6cea07d3-050b-424b-9011-3b7d25034b06)

**Comment by yorkshire-pudding** [comment]
I noticed that this wasn't working for CLI install.  I have added these in as `#hidden` fields on the form.  If there is a suitable source for these to be dropdown and we can also consider validation (i.e. ensure `charset` and `collation` are compatible) then could make configurable in the UI (and CLI) (@izmeez - I'm trying to work it towards the configurable element of your request at #6711)

To make configurable in the CLI, we would need to either move `core/scripts/install.sh` to using individual parameters for DB config (preferred) or stick with db url and add in additional options

Bee can (and will) be adapted as required.

**Comment by yorkshire-pudding** [comment]
> I'm having some difficulty understanding the instructions on testing this. I guess I may just have to experiment with it in conjunction with the PR.
> 
> > EDIT: if you want to test on something like Lando or another system that provides the db settings at server level, you may need to add this line to your `settings.php`:
> > ```
> > if (isset($_SERVER["BACKDROP_SETTINGS"])) unset($_SERVER["BACKDROP_SETTINGS"]);
> > ```

@izmeez - in Lando (and possibly in other docker-based systems) the database settings are never written to `settings.php` and the installer skips the database step unless that line is added to `settings.php`.  You may not need this in your local environment.

**Comment by quicksketch** [comment]
I posted a review to the PR: https://github.com/backdrop/backdrop/pull/4567#pullrequestreview-2361092221

Overall, looking great

**Comment by yorkshire-pudding** [comment]
Thanks @quicksketch - I learned a new bit of syntax and have implemented those changes. The change from `implode` to `serialize` enabled a small simplification which I've also added.

I have tested this through the UI and through `bee` (which uses the `install.sh` script) with a complex password ('P@s$/W0rd#;') and both installed without a problem.

**Comment by quicksketch** [comment]
As @yorkshire-pudding is already advocating for this issue and this is close to completion, this should be in the 1.30.0 milestone (as I just moved it).

**Comment by quicksketch** [comment]
@yorkshire-pudding I tried out this PR and ran into some issues with some edge-case handling.

We have a special check in `DatabaseTasks::connect()` (in `.install.inc`) that detects if `localhost` or `127.0.0.1` is used, and if the connection fails, it tries swapping one for the other. This helped with a common database misconfiguration. This was added in #496.

But that check conflicted with this change because it tried to automatically rewrite `settings.php` when it does the check. And then the installer form tried to rewrite it a second time, and it would result in invalid PHP because the `$database` variable changed its number of lines from the first rewrite.

So to avoid going down the rabbit hole of needing to expand `backdrop_rewrite_settings()`, I pushed a commit to the PR that remove the automatic rewriting. Instead, a message is shown during the installation:

![image](https://github.com/user-attachments/assets/ba01d09a-4e9f-48fe-85e1-bd412165c371)

The entire localhost/127.0.0.1 swapping seems to be less of an issue than it was previously, due to the increase in tools like DDEV and Lando. So removing the automatic attempt fix the issue and request the user set a different value seems okay to me.

@yorkshire-pudding Please take a look at my changes (made directly on your PR) and see if you'd suggest anything different.

**Comment by yorkshire-pudding** [comment]
@quicksketch - your changes make sense.  I have no way to test this 'host' scenario.
@olafgrabienski - thanks for your suggestion, I have clarified where they need to update.

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop/pull/4567 for 1.30.0 that should fully fix the original problem by avoiding needing to URL encode values entirely. This original issue was mostly solved back in 2019 by @jlfranklin (https://github.com/backdrop/backdrop-issues/issues/2231#issuecomment-354702477), but a big thanks to @yorkshire-pudding for finishing the job.

https://github.com/backdrop/backdrop/commit/c53e13d829406bcfcae365754d8345dd57da69f3 by @yorkshire-pudding, @klonos, @izmeez, @quicksketch, @olafgrabienski, and @bugfolder.

This issue is code complete but we need to both update the documentation page at https://docs.backdropcms.org/documentation/database-configuration and write a change record.

**Comment by yorkshire-pudding** [comment]
I'll start the Change record first.

**Comment by yorkshire-pudding** [comment]
I've added a change record at https://docs.backdropcms.org/change-records/database-configuration-can-now-be-stored-in-a-simple-array

I've spotted that @quicksketch has already published a change record for 1.30.0 so I have published this.  Edit if required.
